### ISSUE TITLE
Oracle - completion for other schemas

### ISSUE THREAD

**Issue by p91paul** [dbeaver/dbeaver#5221]
This is a followup of #5208...I did some "research".

Looks like there are 2 types of "completion". One is used when you either specify the schema, or the object you are looking for is in the current schema.
This one just loads the whole object list for that schema then searches on it. This is the best one, as it's fast and fuzzy search fully works. Otherwise, if the object you are looking for does not exist in the current schema, DBeaver triggers the second type of completion: a search on ALL_OBJECTS, most likely using the goto database object code. This kind of search has 2 issues:

- it doesn't use synonyms (as per #5208, that should be fixed in master, albeit possibly breaking part of #5036)
- fuzzy search doesn't work properly, as you search for like '%typedstring%' and only when the result window appears

I believe that there is another solution #5208, that also fixes fuzzy search. The first time completion is triggered, DBeaver issues those queries:

```sql
SELECT  t.OWNER,t.TABLE_NAME as TABLE_NAME,'TABLE' as OBJECT_TYPE,'VALID' as STATUS,t.TABLE_TYPE_OWNER,t.TABLE_TYPE,t.TABLESPACE_NAME,t.PARTITIONED,t.IOT_TYPE,t.IOT_NAME,t.TEMPORARY,t.SECONDARY,t.NESTED,t.NUM_ROWS 
	FROM SYS.DBA_ALL_TABLES t
	WHERE t.OWNER='CURRENT_USER' AND NESTED='NO'
UNION ALL
	SELECT  o.OWNER,o.OBJECT_NAME as TABLE_NAME,'VIEW' as OBJECT_TYPE,o.STATUS,NULL,NULL,NULL,'NO',NULL,NULL,o.TEMPORARY,o.SECONDARY,'NO',0 
	FROM SYS.DBA_OBJECTS o 
	WHERE o.OWNER='CURRENT_USER' AND o.OBJECT_TYPE='VIEW';

SELECT  * FROM SYS.DBA_OBJECTS WHERE OBJECT_TYPE='PACKAGE' AND OWNER='CURRENT_USER'  ORDER BY OBJECT_NAME;

SELECT OWNER, SYNONYM_NAME, MAX(TABLE_OWNER) as TABLE_OWNER, MAX(TABLE_NAME) as TABLE_NAME, MAX(DB_LINK) as DB_LINK, MAX(OBJECT_TYPE) as OBJECT_TYPE FROM (
SELECT S.*, NULL OBJECT_TYPE FROM SYS.DBA_SYNONYMS S WHERE S.OWNER = 'CURRENT_USER'
UNION ALL
SELECT S.*,O.OBJECT_TYPE FROM SYS.DBA_SYNONYMS S, SYS.DBA_OBJECTS O
WHERE S.OWNER = 'CURRENT_USER'
AND O.OWNER=S.TABLE_OWNER AND O.OBJECT_NAME=S.TABLE_NAME
)
GROUP BY OWNER, SYNONYM_NAME
ORDER BY SYNONYM_NAME;
```

This loads, respectively:
- Table and views on the current schema
- Packages on the current schema
- Synonyms in the current schema, or pointing to objects in the current schema

What are the issues here:
- Functions and materialized views are missing (I'm unsure about procedures, AFAIK they can't be used in a SELECT query, but should definitely be loaded in a CALL statement)
- PUBLIC synonyms pointing to objects in another schema

If no object is found, "goto database object" code is triggered, albeit the lack of synonyms in 5.3.4.

I argue that if functions/procedures/materialized views and PUBLIC synonyms are included in the first search, then "goto database object" search during completion is useless. This claim is supported by the fact that your search will find also objects that, despite being visible to the user, cannot be used in the current query unless they are qualified with the owner (because the object is in another schema and is not pointed by a synonym in the current schema or in PUBLIC).
So I believe the proper fix for #5208 is to add PUBLIC synonyms to the first part of completion, not to consider them in the second part (couldn't have I said just that? 😁 ).

**Comment by serge-rider** [comment]
I agreed in general. See fix in 6.0.2.

**Comment by uslss** [comment]
Verified
